neo4j/session_with_context.go

```diff
@@ -545,49 +564,105 @@ func (s *sessionWithContext) getServers(mode idb.AccessMode) func() []string {
-	} else if deadline, ok := ctx.Deadline(); ok {
-		s.log.Debugf(log.Session, s.logId, "connection acquisition user-provided deadline is: %s", deadline)
 	}
 
-	if err := s.resolveHomeDatabase(ctx); err != nil {
-		return nil, errorutil.WrapError(err)
-	}
-	_, err := s.getOrUpdateServers(ctx, mode)
+	serverList, usedHomeDbGuess, err := s.getServerList(ctx, mode)
 	if err != nil {
 		return nil, errorutil.WrapError(err)
 	}
 
-	conn, err := s.pool.Borrow(
-		ctx,
-		s.getServers(mode),
-		timeout != 0,
-		s.config.BoltLogger,
-		livenessCheckTimeout,
-		s.auth)
+	conn, err := s.borrowConnection(ctx, serverList, livenessCheckTimeout)
 	if err != nil {
 		return nil, errorutil.WrapError(err)
 	}
 
+	// Return the connection if SSR is disabled, and we used the home database guess, then resolve as normal.
+	if !conn.IsSsrEnabled() && usedHomeDbGuess {
+		s.pool.Return(ctx, conn)
+
+		serverList, _, err = s.getServerList(ctx, mode)
+		if err != nil {
+			return nil, errorutil.WrapError(err)
+		}
+
+		conn, err = s.borrowConnection(ctx, serverList, livenessCheckTimeout)
+		if err != nil {
+			return nil, errorutil.WrapError(err)
+		}
+	}
+
+	conn.SetPinHomeDatabaseCallback(s.pinHomeDatabaseCallback)
+
 	// Select database on server
 	if s.config.DatabaseName != idb.DefaultDatabase {
-		dbSelector, ok := conn.(idb.DatabaseSelector)
-		if !ok {
-			s.pool.Return(ctx, conn)
-			return nil, &UsageError{Message: "Database does not support multi-database"}
+		if err := s.selectDatabase(ctx, conn); err != nil {
+			return nil, err
 		}
-		dbSelector.SelectDatabase(s.config.DatabaseName)
 	}
 
 	return conn, nil
 }
 
+// applyConnectionTimeout sets a timeout on the context if configured.
+func (s *sessionWithContext) applyConnectionTimeout(ctx context.Context) (context.Context, context.CancelFunc) {
+	timeout := s.driverConfig.ConnectionAcquisitionTimeout
+	if timeout > 0 {
+		ctx, cancel := context.WithTimeout(ctx, timeout)
+		deadline, _ := ctx.Deadline()
+		s.log.Debugf(log.Session, s.logId, "connection acquisition timeout is %s, resolved deadline is: %s", timeout, deadline)
+		return ctx, cancel
+	}
+	if deadline, ok := ctx.Deadline(); ok {
+		s.log.Debugf(log.Session, s.logId, "connection acquisition user-provided deadline is: %s", deadline)
+	}
+	return ctx, nil
+}
+
+// getServerList resolves the server list based on the session configuration.
+// It returns a list of servers, a boolean indicating whether the home database guess was used, and an error if resolution fails.
+func (s *sessionWithContext) getServerList(ctx context.Context, mode idb.AccessMode) ([]string, bool, error) {
+	if s.config.DatabaseName != "" {
+		if err := s.resolveHomeDatabase(ctx); err != nil {
+			return nil, false, err
+		}
+		serverList, err := s.getOrUpdateServers(ctx, mode, false)
+		return serverList, false, err
+	}
+
+	// Use cached home database if available
+	if s.homeDbGuess != "" && s.router.GetTable(s.homeDbGuess) != nil && s.cache.IsEnabled() {
```

Comment:
```suggestion
	if s.homeDbGuess != "" && s.cache.IsEnabled() {
```
`getOrUpdateServers` is going to call `s.router.GetOrUpdateReaders/Writers` regardless which will fetch a RT if none exists. So why bother with this check that could be outdated by the time the router gets to decided whether it needs to fetch a RT or not. Further, the existence of a RT doesn't guarantee that it's valid. The TTL could've expired, there could be no readers/writers left, etc.

Reply:
might need to discuss this code a bit further, without the check TestKit fails along with a unit test for checking this scenario

Reply:
Added onRoutingTableUpdated callback to router for handling this logic rather than relying on GetTable 👍 .